src/components/amino-acid-slider.js

```diff
@@ -125,56 +132,81 @@ class AminoAcidSlider extends Component {
     }
 
     const allelesStyle = {
-      left: `-${this.state.allelesOffset}px`,
-      width: `${this.state.allelesWidth}px`
+      left: `-${this.state.allelesOffset}px`
     }
 
     const acidWidth = 17;
-
-    const AminoAcids = this.props.aminoAcids.split('').map((a, i) =>
-      <AminoAcid type={a} x={acidWidth/2 + (i * (acidWidth * 1.1))} y={3} width={acidWidth} key={i} />
-    )
+    const acidHeight = 15;
+    const fontHeight = 16; // this is the font maximum
+    const acidMargin = 2; // space below amino acids
+    const codonWidth = this.props.showAlleles ? 29 : acidWidth;
+    const codonMargin = codonWidth * 0.1; // space between codons
+    const chainOffset = 9; // space before the acid chain starts
+
+    const location = Math.floor((this.props.aminoAcids.length - 1) * this.props.selectionStart);
+    const AminoAcids = this.props.aminoAcids.split('').map((a, i) => {
+      const codonOffset = chainOffset + i * (codonWidth + codonMargin);
+      const dimmed = this.props.dimUnselected && location !== i;
+      return (
+        <g key={i}>
+          <AminoAcid type={a} x={codonOffset + (codonWidth - acidWidth)/2} y={0} width={acidWidth} dimmed={dimmed} />
+          {
+            this.props.showAlleles &&
+            <Codon dna={this.props.alleles.substring(i * 3, (i + 1) * 3)} x={codonOffset} y={acidHeight + acidMargin + fontHeight} dimmed={dimmed} />
+          }
+        </g>
+      )
+    })
 
     const marks = this.props.marks.map(loc =>
       <rect key={loc} x={acidWidth/2 + (loc * (acidWidth * 1.1)) - 1} y="1" width="19" height="20" style={{fill: "#33F", stroke: "#AAF", strokeWidth: 2}} />
     )
 
+    let svgHeight = acidHeight + acidMargin;
+    if (this.props.showAlleles) {
+      svgHeight += fontHeight;
+    }
+    const svgWidth = (codonWidth + codonMargin) * this.props.aminoAcids.length + chainOffset;
     return (
       <div className={wrapperClass} style={frameStyle} ref={this.wrapperRef}>
-        <div className="alleles" style={allelesStyle} ref={this.alleleStringRef}>
-          <svg width={1000} height={22} viewBox="0 0 1000 22">
-            <path d="M9,12L981,12" style={{stroke: '#AAA', strokeWidth: '2px'}} />
-            { marks }
-            { AminoAcids }
-          </svg>
-        </div>
         <div
           className="selection"
           style={selectStyle}
           ref={this.selectionRef}
           onMouseDown={this.onMouseDown}
           onClick={this.onClick}
         />
+        <div className="amino-acids" style={allelesStyle} ref={this.alleleStringRef}>
+          <svg width={svgWidth} height={svgHeight} viewBox={`0 0 ${svgWidth} ${svgHeight}`}>
```

Comment:
Are these the kind of hardcoded styles you're referencing @lbondaryk? Agreed that it's best to avoid it in general, but I can't think of a cleaner way to calculate these values in CSS.

As for the magic numbers, are you referring to the values above here? I tried to keep them named and documented - perhaps it would be cleaner to declare them as constants at the top of the file.